src/commands/play.rs

```diff
@@ -53,20 +57,68 @@ pub async fn play(
     create_response(&ctx.http, interaction, SEARCHING).await?;
 
-    let enqueue_type = match mode {
-        PlayMode::All | PlayMode::Reverse | PlayMode::Shuffle => EnqueueType::Playlist,
-        _ if url.contains("youtube.com/playlist?list=") => EnqueueType::Playlist,
-        _ if url.starts_with("http") => EnqueueType::Link,
-        _ => EnqueueType::Search,
+    let query_type = if url.contains("youtube.com/playlist?list=") {
+        QueryType::PlaylistLink
+    } else if url.starts_with("http") {
+        QueryType::VideoLink
+    } else {
+        QueryType::Keywords
     };
 
     let call = manager.get(guild_id).unwrap();
 
-    match enqueue_type {
-        EnqueueType::Link => enqueue_song(&call, url.to_string(), true, mode).await,
-        EnqueueType::Search => enqueue_song(&call, url.to_string(), false, mode).await,
-        EnqueueType::Playlist => enqueue_playlist(&call, url, mode).await,
-    };
+    let handler = call.lock().await;
+    let queue_was_empty = handler.queue().is_empty();
+    drop(handler);
+
+    match mode {
+        PlayMode::End => match query_type {
+            QueryType::Keywords | QueryType::VideoLink => {
+                enqueue_song(&call, url.to_string(), query_type).await;
+            }
+            QueryType::PlaylistLink => {
+                enqueue_playlist(&call, url, mode).await;
+            }
+        },
+        PlayMode::Next => match query_type {
+            QueryType::Keywords | QueryType::VideoLink => {
+                enqueue_song(&call, url.to_string(), query_type).await;
+                rotate_tracks(&call, 1).await;
+            }
+            QueryType::PlaylistLink => {
+                if let Some(playlist) = enqueue_playlist(&call, url, mode).await {
+                    rotate_tracks(&call, playlist.len()).await;
+                }
+            }
+        },
+        PlayMode::Jump => match query_type {
+            QueryType::Keywords | QueryType::VideoLink => {
+                enqueue_song(&call, url.to_string(), query_type).await;
+
+                if !queue_was_empty {
+                    rotate_tracks(&call, 1).await;
+                    force_skip_top_track(&call).await;
+                }
+            }
+            QueryType::PlaylistLink => {
+                if let Some(playlist) = enqueue_playlist(&call, url, mode).await {
+                    if !queue_was_empty {
+                        rotate_tracks(&call, playlist.len()).await;
+                        force_skip_top_track(&call).await;
+                    }
+                }
+            }
+        },
+        PlayMode::All | PlayMode::Reverse | PlayMode::Shuffle => match query_type {
+            QueryType::VideoLink | QueryType::PlaylistLink => {
+                enqueue_playlist(&call, url, mode).await;
+            }
+            _ => {
+                edit_response(&ctx.http, interaction, PLAY_ALL_FAILED).await?;
+                return Ok(());
+            }
+        },
```

Comment:
I know the user isn't "supposed to", but if he inputs a single youtube video with this command, the parrot gets stuck.
![image](https://user-images.githubusercontent.com/19473034/154372693-1b5a93a0-5bf5-4f7e-8745-400be8c15dc3.png)
![image](https://user-images.githubusercontent.com/19473034/154372722-1ed9ec30-a8b3-4721-bc9f-3839b71020fa.png)

So maybe I would check for `list=` in the case of a `QueryType::VideoLink`.

Reply:
It's been about 2 years since I wrote my first music bot, and it was in a completely different language but I vaguely remember a flag to make ytdl always return a list of results, even if it only contains a single element. Not saying that it's a better solution, but it won't break Soundcloud playlist support. (This may not even be an option anymore, though.)

Reply:
That is true, but I suppose that they don't face the same issue where we have playlists inside normal video links.

Reply:
Wait, I've got a trick. We're just collecting URLs with the playlist command to then feed back into yt-dlp later, right? Why not use the "webpage_url" attribute instead of just "url" ? The "webpage_url" exists for both playlist and single video queries, and it yields a value that yt-dlp can resolve later.

Reply:
If yt-dlp can resolve that would be cool. Let's check that out @aquelemiguel

Reply:
Awesome suggestion indeed 👀